### ISSUE TITLE
[FSDP] FSDP produces different gradient norms vs DDP, and w/ grad norm clipping creates different training results

### ISSUE THREAD

**Issue by lessw2020** [pytorch/pytorch#88621]
### 🐛 Describe the bug

Running a T5 large on C4 dataset, and using same random seeding, optimizer, lr scheduler, etc, training with DDP vs FSDP NO_SHARD and FULL_SHARD produce different gradient norms and with norm clipping different training outcomes:

![comparison](https://user-images.githubusercontent.com/46302957/200437760-cb9df95e-0aa3-435a-a574-bed5bd4ff14e.png)

The core issue is shown in this logging of the gradient norms:
![comparison_norm](https://user-images.githubusercontent.com/46302957/200437875-f282d44f-b62b-4a25-baf3-a06b2b2ce236.png)

Details on grad norm calculations:
Training was run with l2 norm. 
FSDP uses FSDP clip_grad_norm_()
DDP uses nn.utils.clip_grad_norm_()

code:
~~~
if isinstance(model, FSDP):
                    # grad_norm_before_clip = get_grad_norm_fsdp(model, local_rank)
                    model.clip_grad_norm_(clip_th)
                    grad_norm_after_clip = get_grad_norm_fsdp(model, local_rank, model_sharding_strategy)
                else:
                    # grad_norm_before_clip = get_grad_norm_local(model)
                    nn.utils.clip_grad_norm_(model.parameters(), clip_th)
                    grad_norm_after_clip = get_grad_norm_local(model) 
 ~~~

For logging, this is the code used:
~~~
def get_grad_norm_local(model):
    total_norm = 0.0
    for p in model.parameters():
        local_norm = torch.linalg.vector_norm(p.grad, dtype=p.dtype)
        total_norm += local_norm ** 2
    return total_norm ** 0.5
    
def get_grad_norm_fsdp(model, rank, sharding_strategy=ShardingStrategy.FULL_SHARD):
    local_norm = get_grad_norm_local(model)
    op = torch.distributed.ReduceOp.SUM
    return_norm = local_norm.clone().detach().requires_grad_(False).to(rank)**2
    dist.all_reduce(return_norm, op=op)
    if sharding_strategy == ShardingStrategy.NO_SHARD:
        return_norm = return_norm / world_size
    return return_norm ** 0.5
~~~
 

### Versions

pending...opening issue from partner.

**Comment by cos120** [comment]
Any update?
I train [nanogpt](https://github.com/karpathy/nanoGPT) in FSDP in bf16, I found grad norm is also higher than DDP or zero3 from Deepspeed, this leads loss in FSDP is higher than DDP or zero3.

**Comment by awgu** [comment]
@cos120 There is a chance that this is related to your FSDP wrapping policy in combination with weight tying. Could you share how you wrapped with FSDP? (For example, could you share `print(model)`?)

**Comment by cos120** [comment]
@awgu thanks for your reply.
I use `transformer_auto_wrap_policy` to wrap transformer block in nanoGPT, the wrapped model as shown below.
In nanoGPT, `lm_head` and `wte` are tied together, but they are in same FSDP unit.
I use nightly version, commit id is `6298ac688f8caafe30d71ff2ea2e20fbb32065c7`, I will try 2.1.
versions
```bash
[conda] pytorch-triton            2.1.0+e6216047b8          pypi_0    pypi
[conda] torch                     2.1.0.dev20230731+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230731+cu118          pypi_0    pypi
[conda] torchpippy                0.1.1+cecc4fc            pypi_0    pypi
[conda] torchvision               0.16.0.dev20230731+cu118          pypi_0    pypi
```
```python
FullyShardedDataParallel(
  (_fsdp_wrapped_module): GPT(
    (transformer): ModuleDict(
      (wte): Embedding(50304, 768)
      (wpe): Embedding(1024, 768)
      (drop): Dropout(p=0.0, inplace=False)
      (h): ModuleList(
        (0-11): 12 x FullyShardedDataParallel(
          (_fsdp_wrapped_module): Block(
            (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
            (attn): CausalSelfAttention(
              (c_attn): Linear(in_features=768, out_features=2304, bias=False)
              (c_proj): Linear(in_features=768, out_features=768, bias=False)
              (attn_dropout): Dropout(p=0.0, inplace=False)
              (resid_dropout): Dropout(p=0.0, inplace=False)
            )
            (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
            (mlp): MLP(
              (c_fc): Linear(in_features=768, out_features=3072, bias=False)
              (gelu): GELU(approximate='none')
              (c_proj): Linear(in_features=3072, out_features=768, bias=False)
              (dropout): Dropout(p=0.0, inplace=False)
            )
          )
        )
      )
      (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
    )
    (lm_head): Linear(in_features=768, out_features=50304, bias=False)
  )
)

(Pdb) p model._fsdp_wrapped_module.transformer.wte.weight
Parameter containing:
tensor([ 0.1926,  0.0084, -0.1000,  ..., -0.0139,  0.0759, -0.0374],
       device='cuda:0', requires_grad=True)
(Pdb) p model._fsdp_wrapped_module.lm_head.weight
Parameter containing:
tensor([ 0.1926,  0.0084, -0.1000,  ..., -0.0139,  0.0759, -0.0374],
       device='cuda:0', requires_grad=True)
(Pdb) p model._fsdp_wrapped_module.transformer.wte.weight is model._fsdp_wrapped_module.lm_head.weight
True
(Pdb) 
```

**Comment by cos120** [comment]
I found that this is not releated to FSDP, in nanoGPT, build optimizer depends on dim of parameters. We use `use_orig_param` in FSDP, this leads all parameters are flatten in to 1D tensor, so param group of optimizer is wrong.

Thanks @awgu

**Comment by awgu** [comment]
@cos120 Sounds good! Out of curiosity, which optimizer are you using?
